src/mission_companion.cpp

```diff
@@ -1173,6 +1339,350 @@ bool talk_function::labor_return( npc &p )
     return true;
 }
 
+void talk_function::om_camp_upgrade( npc &comp ){
+    editmap edit;
+    const point omt_pos = ms_to_omt_copy( g->m.getabs( comp.posx(), comp.posy() ) );
+    oter_id &omt_ref = overmap_buffer.ter( omt_pos.x, omt_pos.y, comp.posz() );
+    std::string om_old = omt_ref.id().c_str();
+    if (om_old == "field"){
+        edit.set_mapgen( "faction_base_camp_1", comp.pos());
+    } else {
+        edit.set_mapgen( om_next_upgrade(om_old), comp.pos());
+    }
+
+    tripoint np = tripoint( 5 * SEEX + 5 % SEEX, 4 * SEEY + 5 % SEEY, comp.posz() );
+    np.x = g->u.posx()+1;
+    np.y = g->u.posy();
+
+    tripoint p( 0, 0, comp.posz() );
+    int &x = p.x;
+    int &y = p.y;
+    for( x = 0; x < SEEX * g->m.getmapsize(); x++ ) {
+        for( y = 0; y < SEEY * g->m.getmapsize(); y++ ) {
+            if(g->m.ter(x, y) == ter_id( "t_floor_green" )) {
+                np.x = x;
+                np.y = y;
+            }
+        }
+    }
+    comp.setpos(np);
+    comp.set_destination();
+    g->load_npcs();
+    popup(_("Upgrade completed."));
+}
+
+std::string talk_function::om_upgrade_description( npc &p, std::string bldg ){
+    const recipe *making = &recipe_id( bldg ).obj();
+    if( making == nullptr ) {
+        debugmsg( "No building recipe with id %s found", bldg );
+        return "error";
+    }
+
+    const inventory &total_inv = g->u.crafting_inventory();
+
+    std::vector<std::string> component_print_buffer;
+    int pane = FULL_SCREEN_WIDTH;
+    auto tools = making->requirements().get_folded_tools_list( pane, c_white, total_inv, 1 );
+    auto comps = making->requirements().get_folded_components_list( pane, c_white, total_inv, 1);
+    component_print_buffer.insert( component_print_buffer.end(), tools.begin(), tools.end() );
+    component_print_buffer.insert( component_print_buffer.end(), comps.begin(), comps.end() );
+
+    std::string comp = "";
+    for( auto &elem : component_print_buffer ) {
+        comp = comp + elem + "\n";
+    }
+    std::vector<std::shared_ptr<npc>> npc_list = companion_list( p, "_faction_upgrade_camp" );
+    comp = _("Notes:\n" +
```

Comment:
The `_` only applies to C-strings and is usually not working on generated strings, only on string literals.